### PR TITLE
[WIP]State verification pipeline

### PR DESCRIPTION
**Pull request by forcodedancing**

### CODE REVIEW

File: core/state/statedb.go [hunk 1]
```diff
@@ -1023,6 +1023,23 @@ func (s *StateDB) IntermediateRoot(deleteEmptyObjects bool) common.Hash {
 	return s.StateIntermediateRoot()
 }
 
+func (s *StateDB) AccountsIntermediateWithoutRoot() {
+	for addr := range s.stateObjectsPending {
+		if obj := s.stateObjects[addr]; !obj.deleted {
+			if s.snap != nil && !obj.deleted {
+				s.snapMux.Lock()
```
Comment (realuncle) [inline core/state/statedb.go]: There is no need to lock here.

File: core/state/statedb.go [hunk 2]
```diff
@@ -1023,6 +1023,23 @@ func (s *StateDB) IntermediateRoot(deleteEmptyObjects bool) common.Hash {
 	return s.StateIntermediateRoot()
 }
 
+func (s *StateDB) AccountsIntermediateWithoutRoot() {
```
Comment (realuncle) [inline core/state/statedb.go]: `AccountsIntermediateWithoutRoot ` is not proper any more, need a new name.

File: core/state/snapshot/disklayer.go [hunk 3]
```diff
@@ -59,6 +60,19 @@ func (dl *diskLayer) Verified() bool {
 	return true
 }
 
+func (dl *diskLayer) CorrectAccounts(accounts map[common.Hash][]byte) {
```
Comment (realuncle) [inline core/state/snapshot/disklayer.go]: We need empty implementation for disklayer.

File: core/state/statedb.go [hunk 4]
```diff
@@ -1297,20 +1360,47 @@ func (s *StateDB) Commit(failPostCommitFunc func(), postCommitFuncs ...func() er
...
+		accountRefreshed = make(chan struct{})
 	}
 
 	commmitTrie := func() error {
 		commitErr := func() error {
+			accountData := make(map[common.Hash][]byte)
+			if s.pipeCommit && s.snap != nil {
+				// Due to state verification pipeline, the accounts roots are not updated, leading to the data in the difflayer is not correct
+				// Fix the wrong data here
+				accountData = s.accountDataForDiffLayer()
```
Comment (realuncle) [inline core/state/statedb.go]: `accountData = s.accountDataForDiffLayer()` is needed no matter `s.pipeCommit` `s.snap` is false or nil.

File: core/state/statedb.go [hunk 5]
```diff
@@ -1297,20 +1360,47 @@ func (s *StateDB) Commit(failPostCommitFunc func(), postCommitFuncs ...func() er
...
+	var accountRefreshed chan struct{}
 	if s.snap != nil {
 		diffLayer = &types.DiffLayer{}
 	}
 	if s.pipeCommit {
 		// async commit the MPT
 		verified = make(chan struct{})
 		snapUpdated = make(chan struct{})
+		snapCreated = make(chan common.Hash)
+		accountRefreshed = make(chan struct{})
```
Comment (realuncle) [inline core/state/statedb.go]: `accountRefreshed ` is not needed

File: core/state/snapshot/difflayer.go [hunk 6]
```diff
@@ -286,6 +288,26 @@ func (dl *diffLayer) Verified() bool {
 	}
 }
 
+func (dl *diffLayer) CorrectAccounts(accounts map[common.Hash][]byte) {
```
Comment (forcodedancing) [inline core/state/snapshot/difflayer.go:292]: for debug purpose

File: core/state/statedb.go [hunk 7]
```diff
@@ -1297,20 +1344,45 @@ func (s *StateDB) Commit(failPostCommitFunc func(), postCommitFuncs ...func() er
...
+				snapAccountLock.Lock()
+				s.AccountsIntermediateRoot()
+				snapAccountLock.Unlock()
+				for k, v := range s.snapAccounts {
+					accountData[crypto.Keccak256Hash(k[:])] = v
+				}
+			}
+
 			if s.stateRoot = s.StateIntermediateRoot(); s.fullProcessed && s.expectedRoot != s.stateRoot {
+				fmt.Printf("StateIntermediateRoot - invalid merkle root (remote: %x local: %x)  \n", s.expectedRoot, s.stateRoot)
```
Comment (forcodedancing) [inline core/state/statedb.go]: debug purpose

File: core/state/statedb.go [hunk 8]
```diff
@@ -1023,6 +1023,22 @@ func (s *StateDB) IntermediateRoot(deleteEmptyObjects bool) common.Hash {
 	return s.StateIntermediateRoot()
 }
 
+func (s *StateDB) PopulateSnapAccountAndStorage() {
+	for addr := range s.stateObjectsPending {
+		if obj := s.stateObjects[addr]; !obj.deleted {
+			if s.snap != nil && !obj.deleted {
+				s.populateSnapStorage(obj)
+				s.snapAccounts[obj.address] = snapshot.SlimAccountRLP(obj.data.Nonce, obj.data.Balance, emptyRoot, obj.data.CodeHash)
```
Comment (forcodedancing) [inline core/state/statedb.go]: use empty root here

**Pull request closed without merging**